File: Gauss/src/Matrix/Datatypes/MatrixFloat.java
package Matrix.Datatypes;

public class MatrixFloat extends MatrixDataType<Float> {

    private static final MatrixFloat ONE = new MatrixFloat(1.0f);
    private static final MatrixFloat MINUSONE = new MatrixFloat(-1.0f);
    private static final MatrixFloat ZERO = new MatrixFloat(0.0f);


    private float value;

    public MatrixFloat(int v) {
        setValue(((float) v) / 0x10000); //0x10000 = 2^16
    }

    private MatrixFloat(float v) {
        this.setValue(v);
    }

    @Override
    public MatrixDataType<Float> getZero() {
        return ZERO;
    }

    @Override
    public MatrixDataType<Float> getMinusOne() {
        return MINUSONE;
    }

    @Override
    public MatrixDataType<Float> getOne() {
        return ONE;
    }


    @Override
    public double evaluate() {
        return this.value;
    }

    @Override
    public Float getValue() {
        return this.value;
    }

    @Override
    public void setValue(Float v) {
        this.value = v;
    }

    @Override
    public String toString() {
        return Float.toString(this.value);
    }

    @Override
    public MatrixDataType<Float> add(MatrixDataType<Float> number) {
        return new MatrixFloat(this.getValue() + number.getValue());
    }

    @Override
    public MatrixDataType<Float> subtract(MatrixDataType<Float> number) {
        return new MatrixFloat(this.getValue() - number.getValue());
    }

    @Override
    public MatrixDataType<Float> multiply(MatrixDataType<Float> number) {
        return new MatrixFloat(this.getValue() * number.getValue());
    }

    @Override
    public MatrixDataType<Float> divide(MatrixDataType<Float> number) {
        return new MatrixFloat(this.getValue() / number.getValue());
    }

    @Override
    public MatrixDataType<Float> getInverse() {
        return ONE.divide(this);
    }

    @Override
    public int compareTo(MatrixDataType<Float> number) {
        return Float.compare(this.getValue(), number.getValue());
    }

    @Override
    public MatrixDataType<Float> abs() {
        if (this.compareTo(ZERO) > 0)
            return this;
        else
            return this.multiply(MINUSONE);
    }
}

File: PopulationProtocols/Matrix/Datatypes/MatrixFloat.cs
namespace PopulationProtocols {
    public class MatrixFloat : IMatrixDataType<float> {
        private static readonly MatrixFloat ONE = new MatrixFloat(1.0f);
        private static readonly MatrixFloat MINUSONE = new MatrixFloat(-1.0f);
        private static readonly MatrixFloat ZERO = new MatrixFloat(0.0f);

        private float value;

        private MatrixFloat(float v) {
            value = v;
        }

        public double Evaluate() {
            return (double)value;
        }

        public float GetValue() {
            return value;
        }

        public void SetValue(float v) {
            value = v;
        }

        public override string ToString() {
            return value.ToString();
        }

        public IMatrixDataType<float> Add(IMatrixDataType<float> number) {
            return new MatrixFloat(GetValue() + number.GetValue());
        }

        public IMatrixDataType<float> Subtract(IMatrixDataType<float> number) {
            return new MatrixFloat(GetValue() - number.GetValue());
        }

        public IMatrixDataType<float> Multiply(IMatrixDataType<float> number) {
            return new MatrixFloat(GetValue() * number.GetValue());
        }

        public IMatrixDataType<float> Divide(IMatrixDataType<float> number) {
            return new MatrixFloat(GetValue() / number.GetValue());
        }

        public IMatrixDataType<float> GetInverse() {
            return ONE.Divide(this);
        }

        public int CompareTo(IMatrixDataType<float> number) {
            return GetValue().CompareTo(number.GetValue());
        }

        public IMatrixDataType<float> Abs() {
            if (CompareTo(ZERO) > 0)
                return this;
            else
                return Multiply(MINUSONE);
        }
    }
}